draft-vvv-webtransport-quic.md

```diff
@@ -154,62 +154,70 @@ datagram extension [QUIC-DATAGRAM].
+1. The server sends its header block on stream 0, followed by a FIN indicating
+   that the handshake has been successfully completed.
+
+The QuicTransport header stream payload SHALL be a header block.  A header
+block is a sequence of key-value pairs that are formatted in the following way:
 
 ~~~
- 0                   1                   2                   3
- 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|           Key (16)            |          Length (16)          |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                           Value (*)                         ...
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+QuicTransport Header {
+  Header Name Length (16),
+  Header Name (..),
+  Header Value Length (16),
+  Header Value (..),
+}
 ~~~
-{: #fig-frame title="Client indication format" :}
+{: #fig-frame title="QuicTransport Header Format" :}
 
 The pair includes the following fields:
 
-Key:
-
-: Indicates the field that is being expressed.
-
-Length:
+Name:
 
-: Indicates the length of the value (the length of the key and the length
-  itself are not included).
+: Indicates the name of the field that is being expressed.  The name of the
+  field is case-insensitive, and when encoded on the wire, it MUST consist of
+  ASCII characters in range from `a` to `z`, `0` to `9`, `-`, or `.`.  It MUST
+  begin with an alphanumeric character.
 
 Value:
 
-: The value of the field, the semantics of which are determined by the key.
+: The value of the header field, the semantics of which are determined by the
+  name.
 
-A FIN on the stream 2 SHALL indicate that the message is complete.  The client
-MUST send the entirety of the client indication and a FIN immediately after
-opening the connection.  The server MUST NOT process any application data before
-receiving the entirety of the client indication.  The total length of the client
-indication MUST NOT exceed 65,535 bytes.
+A FIN on the stream 0 SHALL indicate that the message is complete.  The client
+MUST send the entirety of the header block and a FIN immediately after opening
+the connection.  The server MUST NOT process any application data before
+receiving the entirety of the headers.  The total length of a header block MUST
+NOT exceed 1,048,576 bytes; implementations MAY impose additional limit on the
+header block size.
 
-In order to ensure that the user agent can send the client indication
-immediately, the server MUST set `initial_max_streams_uni` transport parameter
-to at least `1`.  The user agent MUST close the connection if the server sets
-`initial_max_streams_uni` to `0`.
+In order to ensure that the user agent can send the headers immediately, the
+server MUST set `initial_max_streams_bidi` transport parameter to at least `1`.
+The user agent MUST close the connection if the server sets
+`initial_max_streams_bidi` to `0`.
```

Comment:
Given the bulleted description of the handshake and this section, it is a little unclear what you expect the server to do. For instance, consider a client cannot send the complete block because the server's initial bidi flow control window is too small. Is the expectation that a Server Application will read data from the QUIC stream (giving flow control credit back to client) and buffer the block in the application layer until the client completes? 

It seems unfortunate if a client can force a server to buffer up to 1,048,576 bytes. Yes the server can hangup, but it would have to wait until the application-imposed limit was reached. It might be healthier for the block to begin with a total length field. That allows a server to early abort.

Reply:
I don't believe anything in the current text of the PR precludes the server from processing headers as they come in, and, say, rejecting based on incorrect origin as soon as it parses the origin.  The requirement is to not _accept_ the headers until they're complete.

We could add a size in front, but this would have a downside of forcing the client to buffer all of the headers to compute the size in advance.

Reply:
The ambiguity here is " MUST NOT process any application data before receiving the entirety of the headers". Its ambiguous because this is an application mapping and its all application data. I think what you mean is that the server MUST not process any received data on streams other than 0, or process any datagrams.

Based on the defined fields so far, I don't see needing a length upfront as a real problem for clients. HTTP/2 and HTTP/3 require a length in HEADERS. What use case really needs such client behaviour?

Reply:
Re ambiguity, thanks for pointing this out! This should be fixed.

Re header size, I agree we should do something about this, but not sure I like explicit size, given that it's redundant otherwise.  HTTP/2 and HTTP/3 do always have size, but that's because neither sends headers on a dedicated stream, so they need header block size for delineation.  I've filed #20 to track this.